src/responder/pendingQueue.ts

```diff
@@ -0,0 +1,309 @@
+     */
+    public toTransactionRequest(): ethers.providers.TransactionRequest {
+        return {
+            chainId: this.request.identifier.chainId,
+            data: this.request.identifier.data,
+            gasLimit: this.request.identifier.gasLimit,
+            nonce: this.nonce,
+            gasPrice: this.currentGasPrice,
+            to: this.request.identifier.to,
+            value: this.request.identifier.value
+        };
+    }
+}
+
+export class GasQueue {
+    /**
+     * Items ordered by an ideal gas prices. Items in the queue
+     * will always be ordered from highest ideal gas price to lowest ideal gas price.
+     * They will also always be ordered by nonce ascending. This ensures that an item
+     * with a high ideal gas price cannot get stuck behind an item with lower ideal
+     * gas price.
+     *
+     * The queue can rearrange the order of transactions, however when it
+     * picks a different item to use the same nonce as a previous item it must increase
+     * current gas price of that item by at least the replacement rate, since a nonce
+     * cannot be replaced with a different transaction without increasing the gas price
+     *  by this amount
+     * @param queueItems The items to put in the queue - must already be ordered correctly
+     * @param emptyNonce
+     *      The next empty nonce. Must equal the last queue item nonce + 1
+     *      if any queue items exists
+     * @param replacementRate
+     *      The amount by which the current gas price of a transaction must be greater than
+     *      an existing transaction if it is to replace it. Expresses as an integer pertange eg
+     *      increase by 13 percent = 13
+     * @param maxQueueDepth The maximum possible number of items that can be put into this queue
+     */
+    public constructor(
+        public readonly queueItems: ReadonlyArray<GasQueueItem>,
+        public readonly emptyNonce: number,
+        public readonly replacementRate: number,
+        public readonly maxQueueDepth: number
+    ) {
+        if (replacementRate < 1) throw new ArgumentError("Replacement rate should be positive.", replacementRate);
+        if (emptyNonce < 0) throw new ArgumentError("Nonce must not be negative.", emptyNonce);
+        if (maxQueueDepth < 1) throw new ArgumentError("Max queue depth must be greater than 0.", maxQueueDepth);
+
+        if (queueItems.length !== 0) {
+            let gasPrice = queueItems[0].idealGasPrice;
+            let nonce = queueItems[0].nonce - 1;
+
+            // check the integrity of the queue
+            for (let index = 0; index < queueItems.length; index++) {
+                const item = queueItems[index];
+                if (item.idealGasPrice.gt(gasPrice)) {
+                    throw new ArgumentError(
+                        "Ideal gas price of queue item was greater than the previous item.",
+                        queueItems
+                    );
+                }
```

Comment:
Yes, this is confusing - there's really nothing I need to check for the first iteration.

I've changed it to check that items.length > 1, then I've started the for loop from index = 1. I havent added any comment though, let me know if you think some are required after seeing the changes.

Reply:
Better, but I still think it is more readable to explicitly read the previous item inside the loop (perhaps calling it `prevItem`, rather than carrying around the `gasPrice` variable from the previous iteration.

Reply:
Done in 3c06558ea7635f828b073b71bdd67d8874cc75fe